include/data_sink.hpp

```diff
@@ -0,0 +1,155 @@
+};
+
+enum class blocking_mode {
+    NonBlocking,
+    Blocking
+};
+
+template<typename T>
+class data_sink : public node<data_sink<T>> {
+public:
+    IN<T>        in;
+    std::size_t  n_samples_consumed = 0;
+    std::size_t  n_samples_max      = -1;
+    int64_t      last_tag_position  = -1;
+    float        sample_rate        = -1.0f;
+
+    static constexpr std::size_t listener_buffer_size = 65536;
+
+    struct poller {
+        std::atomic<bool> finished = false;
+        std::atomic<std::size_t> drop_count = 0;
+        gr::circular_buffer<T> buffer = gr::circular_buffer<T>(listener_buffer_size);
+        decltype(buffer.new_reader()) reader = buffer.new_reader();
+
+        template<typename Handler>
+        [[nodiscard]] bool process(Handler fnc) {
+            const auto available = reader.available();
+            if (available == 0) {
+                return false;
+            }
+
+            const auto read_data = reader.get(available);
+            fnc(read_data);
+            reader.consume(available);
+            return true;
+        }
+    };
+
+private:
+    struct listener {
+        acquisition_mode mode = acquisition_mode::Triggered;
+        std::pair<int64_t, int64_t> window; ///< window of data to return in relation to the matching tag position, e.g. [-2000, 3000] to obtain 2000 presamples and 3000 postsamples
+        std::size_t history_size = 0;
+        bool block = false;
+        int64_t drop_count = 0;
+        std::function<bool(fair::graph::tag_t::map_type)> trigger_predicate;
+        gr::circular_buffer<T> buffer;
+        std::optional<std::size_t> pending; ///< number of samples expected due to a previous trigger
+        std::function<void(std::span<const T>)> callback; // TODO we might want to pass back stats here like drop_count
+        std::weak_ptr<poller> polling_handler;
+    };
+
+    struct {
+        std::atomic<bool> dirty = false;
+        std::mutex mutex;
+        std::vector<listener> list;
+    } pending_listeners;
+
+    std::vector<listener> listeners;
+
+public:
+    std::shared_ptr<poller> get_streaming_poller(blocking_mode block = blocking_mode::NonBlocking) {
+        auto handler = std::make_shared<poller>();
+        pending_listeners.list.push_back({
+            .mode = acquisition_mode::Continuous,
+            .block = block == blocking_mode::Blocking,
+            .buffer = gr::circular_buffer<T>(0),
+            .polling_handler = handler
+        });
+        pending_listeners.dirty = true;
+        return handler;
+    }
+
+    template<typename Callback>
+    void register_streaming_callback(Callback callback) {
+        std::lock_guard lg(pending_listeners.mutex);
+        pending_listeners.list.push_back({
+            .mode = acquisition_mode::Continuous,
+            .buffer = gr::circular_buffer<T>(0),
+            .callback = std::move(callback)
+        });
+        pending_listeners.dirty = true;
+    }
+
+    [[nodiscard]] work_return_t work() {
```

Comment:
> Rather than implementing the full `work()` function it might be easier to just implement the `process_bulk(std::span<const T> in) {...}` function. This way you can re-use a lot of the underlying tag and error reporting mechanics.

I looked at process_bulk, but my problem was that I want don't necessarily want to process all the available data, and I didn't see a way to report that.

Reply:
I added that usecase to the work function refactoring issue #81

Reply:
Thinking about it, I wonder if there's a good reason not to process everythin, dropping/blocking more aggressively.